File: 4284-smallest-stable-index-i/smallest-stable-index-i.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        n = len(nums)
        suf = [0] * n
        suf[-1] = nums[-1]

        for i in range(n - 2, -1, -1):
            suf[i] = min(suf[i + 1], nums[i])

        mx = 0
        for i, x in enumerate(nums):
            mx = max(mx, x)
            if mx - suf[i] <= k:
                return i

        return -1

File: 4284-smallest-stable-index-i/smallest-stable-index-i.js
const firstStableIndex = (nums, k, pref = -1) => {
    const suf = [];
    nums.reduceRight((a, c, i) => suf[i] = Math.min(a, c), Infinity);
    return nums.findIndex((c, i) => (pref = Math.max(pref, c)) - suf[i] <= k);
};